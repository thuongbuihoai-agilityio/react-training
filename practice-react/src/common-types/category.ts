export interface Categories {
  id: string;
  name: string;
  images: {
    src: string;
    alt: string;
  };
}

export interface CategoryContext {
  searchValue: {} | undefined;
  setSearchValue: Function;
  categories: Categories[];
  setCategories: Function;
}