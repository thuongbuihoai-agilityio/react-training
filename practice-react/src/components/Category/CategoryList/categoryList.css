.categories {
  padding: 120px 0 0 125px;
}

.categories__title {
  margin-bottom: 69px;
  font-weight: 600;
  font-size: 24px;
  line-height: 36px;
  text-transform: capitalize;
  color: #2B463C;
}

.categoryList--select {
  display: flex;
  margin-bottom: 246px;
}

.categoryList__list {
  display: flex;
  flex-direction: column;
  margin-right: 55px;
  background: #FFFFFF;
  border-radius: 5px;
  width: 235px;
  box-shadow: -50px 50px 200px #F2F2F2;
}

.categoryList__checkbox {
  margin: 0 0 23px 73px;
  font-weight: 400;
  font-size: 24px;
  line-height: 36px;
  color: #2B463C;
}

.categoryList__image {
  margin: 32px 0 60px;
  text-align: center;
}