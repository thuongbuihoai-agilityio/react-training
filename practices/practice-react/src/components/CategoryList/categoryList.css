.categories {
  padding: 120px 0 0 125px;
}

.categories-title {
  margin-bottom: 69px;
  font-weight: 600;
  font-size: 24px;
  line-height: 36px;
  text-transform: capitalize;
  color: var(--text-primary);
}

.category-list-row {
  display: flex;
  margin-bottom: 246px;
}

.category-list-card {
  display: flex;
  flex-direction: column;
  margin-right: 55px;
  background: var(--light);
  border-radius: 5px;
  width: 235px;
  box-shadow: -50px 50px 200px var(--box--light);
}

.category-list-checkbox {
  margin: 0 0 23px 73px;
  font-weight: 400;
  font-size: 24px;
  line-height: 36px;
  color: var(--text-primary);
}