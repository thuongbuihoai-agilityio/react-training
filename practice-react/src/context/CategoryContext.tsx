import { Categories, CategoryContext } from "@common-types/category";
import { createContext, useMemo, useState } from "react";

const CategoriesContext = createContext<CategoryContext>({} as CategoryContext);
const CategoriesProvider: React.FC<{ children: JSX.Element[] | JSX.Element }> = ({
  children,
}) => {
  const [searchValue, setSearchValue] = useState<{}>();
  const [categories, setCategories] = useState<Categories[]>([]);

  const value = useMemo(
    () => ({
      searchValue,
      setSearchValue,
      categories,
      setCategories,
    }),
    [searchValue]
  );

  return <CategoriesContext.Provider value={value}>{children}</CategoriesContext.Provider>;
};

export { CategoriesProvider, CategoriesContext };
