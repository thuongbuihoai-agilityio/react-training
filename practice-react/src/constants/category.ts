import { CategoryType } from "@common-types/categoryList";
import iconBeverages from "@assets/images/icons/beverages.png";
import iconDairy from "@assets/images/icons/dairy.png";
import iconMeat from "@assets/images/icons/meat.png";
import iconSpicesHerbs from "@assets/images/icons/spices-herbs.png";
import iconStationery from "@assets/images/icons/stationery.png";
import iconCostmetic from "@assets/images/icons/costmetic.png";

const CATEGORY: CategoryType = {
  key: "beverages",
  src: iconBeverages,
  text: "Beverages",
  alt: "This is icon beverages",
};

const CATEGORY_LIST: CategoryType[] = [
  {
    key: "beverages",
    src: iconBeverages,
    text: "Beverages",
    alt: "This is icon beverages",
  },
  {
    key: "dairy",
    src: iconDairy,
    text: "Dairy",
    alt: "This is icon dairy",
  },
  {
    key: "meat",
    src: iconMeat,
    text: "Meat",
    alt: "This is icon meat",
  },
  {
    key: "spices-herbs",
    src: iconSpicesHerbs,
    text: "Spices & Herbs",
    alt: "This is icon spices-herbs",
  },
  {
    key: "stationery",
    src: iconStationery,
    text: "Stationery",
    alt: "This is icon stationery",
  },
  {
    key: "costmetic",
    src: iconCostmetic,
    text: "Costmetic",
    alt: "This is icon costmetic",
  },
];

export { CATEGORY, CATEGORY_LIST };
