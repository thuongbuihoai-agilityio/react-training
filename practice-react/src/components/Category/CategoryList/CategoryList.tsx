import useSWR from "swr";
import React, { useContext, useEffect } from "react";
import Checkbox from "@components/common/Checkbox/Checkbox";
import { Categories } from "@common-types/category";
import { Link } from "react-router-dom";
import { getData } from "@helpers/fetchApi";
import { CATEGORIES_URL } from "@constants/url";
import { CategoriesContext } from "@context/CategoryContext";
import "./categoryList.css";

interface CategoryProps {
  isCheckbox?: boolean;
  isSelect?: boolean;
  type: string;
}

const CategoryList: React.FC<CategoryProps> = ({
  isCheckbox,
  isSelect,
  type,
}) => {
  // fetch data with useSWR
  const { data } = useSWR(CATEGORIES_URL, getData<Categories[]>);
  const { setSearchValue, setCategories } = useContext(CategoriesContext);

  useEffect(() => {
    setCategories(data);
  }, [data]);

  const handleSearch = (id: string) => (e: React.MouseEvent<HTMLElement>) => {
    // get current categoryId
    const categoryId = { categoryId: e.currentTarget.dataset.index };
    setSearchValue?.(categoryId);
  };

  // handle search default
  const handleDefaultCategory = () => {
    const categoryId = "";
    setSearchValue?.(categoryId);
  };

  let className = "";
  switch (type) {
    case "select":
      className = "categoryList--select";
      break;
    case "checkbox":
      className = "categoryList--checkbox";
      break;
    default:
      break;
  }

  const renderCategoryList = (data: Categories[]) => {
    return data?.map((category: Categories) => (
      <div key={category.id}>
        {isSelect && (
          <div className="categoryList__list">
            <figure className="categoryList__image">
              <img
                className="categoryList__images"
                src={category.images.src}
                alt={category.images.alt}
              />
            </figure>
            <Link data-index={category.id} className="categoryList__text" to="/products" onClick={handleSearch(category.id)}>
              <p>{category.name}</p>
            </Link>
          </div>
        )}
        {isCheckbox && (
          <div className="categoryList__checkbox">
            <Checkbox value={category.name} />
          </div>
        )}
      </div>
    ));
  };

  return (
    <div data-testid="category-list" className={className}>
      {renderCategoryList(data as [])}
    </div>
  );
};

export default CategoryList;
