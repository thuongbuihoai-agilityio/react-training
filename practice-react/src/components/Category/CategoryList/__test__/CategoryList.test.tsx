import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import "@testing-library/jest-dom";
import CategoryList from "../CategoryList";
import { CategoryContext } from "@common-types/category";
import { CategoriesContext } from "@context/CategoryContext";

const contextProductMock: CategoryContext = {
  searchValue: {},
  setSearchValue: () => {},
  categories: [],
  setCategories: () => {},
};

describe("CategoryList component", () => {
  test("should render categoryList - type select component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <CategoriesContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <CategoryList type="select" isSelect={true} />
        </Router>
      </CategoriesContext.Provider>
    );
    expect(getByTestId("category-list")).toBeInTheDocument();
  });

  test("should render categoryList - type checkbox component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <CategoriesContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <CategoryList type="checkbox" isCheckbox={true} />
        </Router>
      </CategoriesContext.Provider>
    );
    expect(getByTestId("category-list")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <CategoriesContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <CategoryList type="select" isSelect={true} />
        </Router>
      </CategoriesContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
