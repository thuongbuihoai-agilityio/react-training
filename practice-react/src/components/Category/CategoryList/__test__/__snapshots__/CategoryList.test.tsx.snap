// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="categoryList--select"
    data-testid="category-list"
  />
</DocumentFragment>
`;
